tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/GreedyHeuristicJoinOrderAlgorithm.java

```diff
@@ -57,17 +54,69 @@ public FoundJoinOrder findBestOrder(LogicalPlan plan, LogicalPlan.QueryBlock blo
     JoinEdge bestPair;
 
     while (remainRelations.size() > 1) {
+      Set<LogicalNode> checkingRelations = new LinkedHashSet<LogicalNode>();
+
+      for (LogicalNode relation : remainRelations) {
+        Collection <String> relationStrings = PlannerUtil.getRelationLineageWithinQueryBlock(plan, relation);
+        List<JoinEdge> joinEdges = new ArrayList<JoinEdge>();
+        String relationCollection = TUtil.collectionToString(relationStrings, ",");
+        List<JoinEdge> joinEdgesForGiven = joinGraph.getIncomingEdges(relationCollection);
+        if (joinEdgesForGiven != null) {
+          joinEdges.addAll(joinEdgesForGiven);
+        }
+        for (String relationString: relationStrings) {
+          joinEdgesForGiven = joinGraph.getIncomingEdges(relationString);
+          if (joinEdgesForGiven != null) {
+            joinEdges.addAll(joinEdgesForGiven);
+          }
+        }
+
+        // check if the relation is the last piece of outer join
+        boolean endInnerRelation = false;
+        for (JoinEdge joinEdge: joinEdges) {
```

Comment:
Here, joinEdges seem to be sorted in the order of their occurrence in the user query.
So, I wonder how we can guarantee that joinEdges are always sorted in that order.